kytos/core/auth.py

```diff
@@ -44,6 +55,108 @@ def wrapper(*args, **kwargs):
 
 
+@for_all_methods(
+    retries,
+    stop=stop_after_attempt(
+        int(os.environ.get("MONGO_AUTO_RETRY_STOP_AFTER_ATTEMPT", 3))
+    ),
+    wait=wait_random(
+        min=int(os.environ.get("MONGO_AUTO_RETRY_WAIT_RANDOM_MIN", 0.1)),
+        max=int(os.environ.get("MONGO_AUTO_RETRY_WAIT_RANDOM_MAX", 1)),
+    ),
+    before_sleep=before_sleep,
+    retry=retry_if_exception_type((AutoReconnect,)),
+)
+class UserController:
+    """UserController"""
+
+    # pylint: disable=unnecessary-lambda
+    def __init__(self, get_mongo=lambda: Mongo()):
+        self.mongo = get_mongo()
+        self.db_client = self.mongo.client
+        self.db = self.db_client[self.mongo.db_name]
+
+    def bootstrap_indexes(self):
+        """Bootstrap all users related indexes."""
+        index_tuples = [
+            ("users", [("username", pymongo.ASCENDING)], {"unique": True}),
+        ]
+        for collection, keys, kwargs in index_tuples:
+            if self.mongo.bootstrap_index(collection, keys, **kwargs):
+                LOG.info(
+                    f"Created DB index {keys}, collection: {collection}"
+                )
+
+    def create_user(self, user_data: dict) -> pymongo.InsertOne:
+        """Create user to database"""
+        try:
+            utc_now = datetime.datetime.utcnow()
+            result = self.db.users.insert_one(UserDoc(**{
+                **user_data,
+                **{"inserted_at": utc_now},
+                **{"updated_at": utc_now}
+            }).dict())
+        except (ValidationError, DuplicateKeyError) as e:
+            raise e
+        return result
+
+    def delete_user(self, username: str) -> dict:
+        """Delete user from database"""
+        utc_now = datetime.datetime.utcnow()
+        result = self.db.users.find_one_and_update(
+            {"username": username},
+            {"$set": {
+                "state": "inactive",
+                "deleted_at": utc_now,
+            }},
+            return_document=ReturnDocument.AFTER,
+        )
+        return result
+
+    def update_user(self, username: str, data: dict) -> dict:
+        """Update user from database"""
+        utc_now = datetime.datetime.utcnow()
+        try:
+            result = self.db.users.find_one_and_update(
+                {"username": username},
+                {
+                    "$set": UserDocUpdate(**{
+                        **data,
+                        **{"updated_at": utc_now}
+                    }).dict(exclude_none=True)
+                },
+                return_document=ReturnDocument.AFTER
+            )
+        except ValidationError as e:
+            raise e
+        return result
+
+    def get_user(self, username: str) -> dict:
+        """Return a user information from database"""
+        data = self.db.users.aggregate([
+            {"$match": {"username": username}},
+            {"$project": UserDoc.projection()},
+            {"$limit": 1}
+        ])
+        return data.next()
```

Comment:
Calling `.next()` on this generator when it returns no user will crash since it'll raise `StopIteration` (same thing on `get_user_now_pw` method below. Let's consume this generator safely by mapping it into a list, and return the head of the list if the list isn't empty. @Alopalao make sure to also explore the endpoint locally, let's also add a unit test covering this corner case when a user isn't found.

Reply:
Done.